### ISSUE TITLE
scrollBehavior setting doesn't work

### ISSUE THREAD

**Issue by hecateball** [nuxt/framework#1707]
### Environment

------------------------------
- Operating System: `Darwin`
- Node Version:     `v16.11.0`
- Nuxt Version:     `3.0.0-27267528.292b524`
- Package Manager:  `npm@8.0.0`
- Bundler:          `Vite`
- User Config:      `buildModules`, `css`, `meta`
- Runtime Modules:  `-`
- Build Modules:    `nuxt-windicss@^2.0.3`
------------------------------

### Describe the bug

`scrollBehavior` setting (which is available in Nuxt2) doesn't work.
I tried both nuxt.config.ts `router.scrollBehavior` and putting `router.scrollBehavior.js` in project route but they have no effect.

I couldn't find any document on scrollBehavior at Nuxt3 docs website... Is there any way to make it work? 

### Reproduction

nuxt.config.ts

```
import { defineNuxtConfig } from 'nuxt3'

export default defineNuxtConfig({
  // ......
  router: {
    scrollBehavior: () => ({ x: 0, y: 0 }),
  },
})
```

~/router.scrollBehavior.js
```
export default () => {
  console.log('here') // never called
  return { x: 0, y: 0 }
}

```

**Comment by danielroe** [comment]
You can set this in a plugin:
```js
export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.$router.options.scrollBehavior = () => {
    return { left: 0, top: 0 }
  }
})
```

**Comment by hecateball** [comment]
@danielroe Thank you!
Is that  a standard way to configure router in Nuxt 3?

**Comment by dergunovs** [comment]
> You can set this in a plugin:
> 
> ```js
> export default defineNuxtPlugin((nuxtApp) => {
>   nuxtApp.$router.options.scrollBehavior = () => {
>     return { left: 0, top: 0 }
>   }
> })
> ```

This fix works not so smooth. White blink spoils the impression.

**Comment by K-Honkawa** [comment]
I'm sorry.
I have reopened...

**Comment by Gyurmatag** [comment]
The plugin solution is not working smooth. Any idea when will it be fixed?

**Comment by madsh93** [comment]
Another thing I noticed is if you add anything with router.push it will scroll to top, even though it may not be what you want to do.

```js
    router.push({
      query: {
        ...route.query,
        equipment: reAddedEquipment.map((x) => x.id)
      }
    })
```

**Comment by danielroe** [comment]
There is a PR in progress: https://github.com/nuxt/framework/pull/3485

**Comment by danielroe** [comment]
This is now possible with https://github.com/nuxt/framework/pull/3485. See https://v3.nuxtjs.org/docs/directory-structure/pages#router-options for documentation.

**Comment by li9269391** [comment]
> This is now possible with #3485. See https://v3.nuxtjs.org/docs/directory-structure/pages#router-options for documentation.
Configuration has been attempted but not effective. Has anyone succeeded？

**Comment by li9269391** [comment]
> This is now possible with #3485. See https://v3.nuxtjs.org/docs/directory-structure/pages#router-options for documentation.

Configuration has been attempted but not effective. Has anyone succeeded？

**Comment by KevinNTH** [comment]
Yes @li9269391 
Using the recent release candidate:

```js
// app/router.options.js
export default {
  scrollBehavior() {
    return { top: 0 }
  },
}
```

**Comment by Devaiser** [comment]
> 

Is that all the code that is in this file? It just doesn’t work for me and I don’t quite understand how to properly configure the router.options

**Comment by KevinNTH** [comment]
@Devaiser yes, that's it for me. I just need to extend the `scrollBehavior` router option.

nuxt3 version: `^3.0.0-rc.0-27508091.78fcbcf`

**Comment by TJacoob** [comment]
If, like me, you were not able to get this working. The problem was that I was placing the `router.options` file at the root of the project, instead of the `/app` folder. 

It would be nice to have a full example of this in the docs, as right now the code is spread across issues.

**Comment by KevinNTH** [comment]
@TJacoob I saw it is actually mentioned in the docs:
![image](https://user-images.githubusercontent.com/1632391/164949753-771e4b21-7eb7-4293-9608-db1a3dc378c9.png)

**Comment by Devaiser** [comment]
As already mentioned above, this setting works rather strangely, the page first scrolls up, and only then switches to a new route

**Comment by li9269391** [comment]
> > 
> 
> @KevinNTH Is that all the code that is in this file? It just doesn’t work for me and I don’t quite understand how to properly configure the router.options

<img width="319" alt="image" src="https://user-images.githubusercontent.com/10287877/165116228-f2811578-39c4-466a-80cb-b6b00cbebea6.png">

good,It's working

**Comment by TJacoob** [comment]
> @TJacoob I saw it is actually mentioned in the docs

I saw that too but didn't think too much of it. The `app` folder is not mentioned in the directory structure of the guide, so I **wrongly** assumed that the `app` prefix meant the root folder of the app :sweat_smile: 

I understand that the docs are improving as we speak, this is meant to help others that might be making the same mistake I did.

**Comment by aacassandra** [comment]
> > Yes @li9269391 Using the recent release candidate:
> > ```js
> > // app/router.options.js
> > export default {
> >   scrollBehavior() {
> >     return { top: 0 }
> >   },
> > }
> > ```
> 
> <img alt="image" width="295" src="https://user-images.githubusercontent.com/10287877/165116752-24bff580-6f21-485c-b25f-a80cb0cf42da.png">
> 
> It's working

yes it works!

but... for the first change hehe
second page, and so on doesn't work

**Comment by adamchipperfield** [comment]
Agree that this isn't clear. Is there a way to make the location of this file configurable? Not too much a fan of having one file in `app` if I can help it. Not a huge deal if not.

**Comment by klyuk88** [comment]
The transition is not smooth, the page flies up, then the transition is made. ((

**Comment by tewshi** [comment]
For Nuxt 3 RC 4 that I use, 

```typescript
return { left: 0, top: 0 }
```

not 
```typescript
return { x: 0, y: 0 }
```

**Comment by jonmes** [comment]
this might work


no need to configure _`nuxt.config.ts`_

`/plugins/router.scrollBehaviour.js`
```
export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.$router.options.scrollBehavior = async (to, from, savedPosition) => {
    if (savedPosition) {
      return savedPosition;
    }

    const findEl = async (hash, x = 0) => {
      return (
        document.querySelector(hash) ||
        new Promise((resolve) => {
          if (x > 0) {
            return resolve(document.querySelector("#app"));
          }
          setTimeout(() => {
            resolve(findEl(hash, 1));
          }, 100);
        })
      );
    };

    if (to.hash) {
      const el = await findEl(to.hash);

      if ("scrollBehavior" in document.documentElement.style) {
        console.log("hash path hit scroll to");
        // return el.scrollTo({ top: el.offsetTop, behavior: "smooth" });
        return window.scrollTo({ top: el.offsetTop, behavior: "smooth" });
      } else {
        return window.scrollTo(0, el.offsetTop);
      }
    }
    return { left: 0, top: 0, behaviour: "smooth" };
  };
});

```

**Comment by GonzaloHirschToptal** [comment]
> this might work
> 
> no need to configure _`nuxt.config.ts`_
> 
> `/plugins/router.scrollBehaviour.js`
> 
> ```
> export default defineNuxtPlugin((nuxtApp) => {
>   nuxtApp.$router.options.scrollBehavior = async (to, from, savedPosition) => {
>     if (savedPosition) {
>       return savedPosition;
>     }
> 
>     const findEl = async (hash, x = 0) => {
>       return (
>         document.querySelector(hash) ||
>         new Promise((resolve) => {
>           if (x > 0) {
>             return resolve(document.querySelector("#app"));
>           }
>           setTimeout(() => {
>             resolve(findEl(hash, 1));
>           }, 100);
>         })
>       );
>     };
> 
>     if (to.hash) {
>       const el = await findEl(to.hash);
> 
>       if ("scrollBehavior" in document.documentElement.style) {
>         console.log("hash path hit scroll to");
>         // return el.scrollTo({ top: el.offsetTop, behavior: "smooth" });
>         return window.scrollTo({ top: el.offsetTop, behavior: "smooth" });
>       } else {
>         return window.scrollTo(0, el.offsetTop);
>       }
>     }
>     return { left: 0, top: 0, behaviour: "smooth" };
>   };
> });
> ```

This solution works for me, but I experience an error in the following scenario. 

I have a navigation containing links to anchors on a homepage and a single link to a different page. If the homepage was rendered at least once, it works perfectly when navigating from other pages. Now, if i first load any page other than the homepage, when clicking on the nav links to the anchors on the homepage, it fails because the element it tries to find returns null, thus having no offsetTop and causing an exception.

Do I need to configure anything else to account for that?

**Comment by GonzaloHirschToptal** [comment]
> > this might work
> > no need to configure _`nuxt.config.ts`_
> > `/plugins/router.scrollBehaviour.js`
> > ```
> > export default defineNuxtPlugin((nuxtApp) => {
> >   nuxtApp.$router.options.scrollBehavior = async (to, from, savedPosition) => {
> >     if (savedPosition) {
> >       return savedPosition;
> >     }
> > 
> >     const findEl = async (hash, x = 0) => {
> >       return (
> >         document.querySelector(hash) ||
> >         new Promise((resolve) => {
> >           if (x > 0) {
> >             return resolve(document.querySelector("#app"));
> >           }
> >           setTimeout(() => {
> >             resolve(findEl(hash, 1));
> >           }, 100);
> >         })
> >       );
> >     };
> > 
> >     if (to.hash) {
> >       const el = await findEl(to.hash);
> > 
> >       if ("scrollBehavior" in document.documentElement.style) {
> >         console.log("hash path hit scroll to");
> >         // return el.scrollTo({ top: el.offsetTop, behavior: "smooth" });
> >         return window.scrollTo({ top: el.offsetTop, behavior: "smooth" });
> >       } else {
> >         return window.scrollTo(0, el.offsetTop);
> >       }
> >     }
> >     return { left: 0, top: 0, behaviour: "smooth" };
> >   };
> > });
> > ```
> 
> This solution works for me, but I experience an error in the following scenario.
> 
> I have a navigation containing links to anchors on a homepage and a single link to a different page. If the homepage was rendered at least once, it works perfectly when navigating from other pages. Now, if i first load any page other than the homepage, when clicking on the nav links to the anchors on the homepage, it fails because the element it tries to find returns null, thus having no offsetTop and causing an exception.
> 
> Do I need to configure anything else to account for that?

I have just found out that if I increase the timeout from 100 to 300, it gives it enough time so that it can properly find it. Definitely not an elegant solution, but it works for now.

**Comment by jhull** [comment]
I could use the delay too.

The documented approach here: https://nuxtjs.org/docs/configuration-glossary/configuration-router#scrollbehavior did not work for me.

The only thing that worked was the above solution, with naming the file `router.options.js` placing it in the `app` directory and then changing to `top: 0, left: 0`.

But there is still that flash of the current page scrolling up first, really quick, and then it goes to the new page.

**Comment by alexanderhorner** [comment]
Yes, please reopen this. The scroll behavior is very jarring. On click it immediately jumps to the top, and then the new page loads.

The correct behavior should be seamless. When the new page renders, it should be scrolled to the correct position.

**Comment by danielroe** [comment]
@alexanderhorner This issue is closed because it is implemented (it is configurable). There are [several other open issues](https://github.com/nuxt/framework/search?q=scrollbehavior+is%3Aissue&type=issues) that describe what you want.

**Comment by ricky11** [comment]
The nuxt3 docs/guide make no mention of `app` folder in the structure, and then some where hidden in the docs there is a mention of a app/router.options.js file that needs to be configured.  🤔 One would have to parse github issues to really understand what is going on. Why not in nuxt.config.js under router options?

in vue 2.x all i would need to do in router.js would be :
`  scrollBehavior(to, from,savedPosition) {
    return {x:0,y:0}
  },`
  
  this would scroll the next route to the top (like most ssr apps) without any jarring animaions or flicker.
  
  The current way mentioned here will scrop to top of the current page and then shift to the next route. (rc11.)
  
  The only options that worked but still not the way it is 'suppose to be ' is to use a plugin with the `page:finish` nuxt hook 
  
  `export default defineNuxtPlugin((nuxtApp) => {
nuxtApp.hook('page:finish', () => {
      window.scrollTo(0, 0);
    })
   })`
   
   This will load the next route, and then scroll up (also kinda strange)
   
   Nuxt3 needs a cookbook.

**Comment by madsh93** [comment]
It should be built in as a default that it scrolls to top on page change. That is how regular sites work and what users/developers expect. Nuxt goes a long way to improve both UX/DX but seems to have missed this one.

**Comment by ricky11** [comment]
This is actually a vue-router issue, but since Nuxt is amazing and `<nuxt-link>` and `generateTo` are wrappers of vue Router methods, they should be "intuitive" on the way the page routing UX should be like.

If you have solved this issue in nuxt3, please do share your reproduction.. all my tests either have a visible page scroll up before or after the route change.. which is not what most people want.

**Comment by Ischafak** [comment]
> scrollBehavior(to, from,savedPosition) { return {x:0,y:0} }

Hi, i tried this command and it doesnt work. Please 
I created a scroll.js inside plugins folder
and my code inside of the file is:

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.hook("page:finish", () => {
    window.scrollTo(0, 0)
  })
})

I am changing page with nuxtlink but when i change page scroll stays same position

**Comment by alexanderhorner** [comment]
Do it the official way: [like this](https://stackoverflow.com/a/73893669/10836492) .

**Comment by Ischafak** [comment]
@alexanderhorner thank you so much you made my day.
Can i ask you 1 more thing
Even if user try to go same page the scroll goes to top of the page but in nuxt 2 it doesnt go top for same scenario. 
Which one is correct and can we do something like nuxt 2 in nuxt3?

**Comment by aminurislamarnob** [comment]
**Solution for Nuxt3 (RC):**
You need to add this code inside ~/app/router.options.ts
`
import type { RouterOptions } from '@nuxt/schema'
export default <RouterOptions> {
  scrollBehavior(to, from, savedPosition) {
    if (savedPosition) {
      return savedPosition
    } else {
      return { 
        top: 0,
        behavior: 'smooth',
      }
    }
  }
}`

**Comment by ricky11** [comment]
> 

This - scrolls to top BEFORE route navigation, hence its better to remove smooth behaviour here.
This would probably only make sense if your are routing to a #  on the same page at the top.
